tests/test_reweighting.py

```diff
@@ -0,0 +1,407 @@
+            u_s = U_s / (BOLTZ * temperature)
+
+            return u_s
+
+        return batched_u_fxn
+
+    batched_u_0 = make_batched_u_fxn(lam=0.0)
+    batched_u_1 = make_batched_u_fxn(lam=1.0)
+
+    return samples_0, samples_1, batched_u_0, batched_u_1, ref_params, ref_delta_f
+
+
+def test_endpoint_reweighting_ahfe():
+    """on made-up inputs of the right shape,
+    check that derivative of an absolute hydration free energy w.r.t .ligand nonbonded parameters can be computed using
+    custom ops
+    """
+    onp.random.seed(2022)
+
+    fake_samples_0, fake_samples_1, batched_u_0, batched_u_1, ref_params, ref_delta_f = make_ahfe_test_system()
+
+    estimate_delta_f = construct_endpoint_reweighting_estimator(
+        fake_samples_0, fake_samples_1, batched_u_0, batched_u_1, ref_params, ref_delta_f
+    )
+
+    v, g = value_and_grad(estimate_delta_f)(ref_params)
+
+    assert np.isfinite(v)
+    assert v == ref_delta_f
+    assert np.isfinite(g).all()
+    assert (g != 0).any()
+    assert g.shape == ref_params.shape
+    # assert anything_about_direction_of_g  # not expected because the "sample" arrays are made up
+
+    # expect different estimate when evaluated on slightly different parameters
+    params_prime = ref_params + 0.01 * onp.random.randn(*ref_params.shape)
+    v_prime, g_prime = value_and_grad(estimate_delta_f)(params_prime)
+    assert v_prime != v
+    assert (g_prime != g).any()
+    assert np.isfinite(v_prime)
+    assert np.isfinite(g_prime).all()
+
+
+def test_mixture_reweighting_ahfe():
+    """on made-up inputs of the right shape,
+    check that derivative of an absolute hydration free energy w.r.t .ligand nonbonded parameters can be computed using
+    custom ops
+    """
+    onp.random.seed(2022)
+
+    _samples_0, _samples_1, batched_u_0, batched_u_1, ref_params, ref_delta_f = make_ahfe_test_system()
+    fake_samples = _samples_1 + _samples_1
+    fake_log_weights = onp.random.randn(len(fake_samples))
+
+    estimate_delta_f = construct_mixture_reweighting_estimator(fake_samples, fake_log_weights, batched_u_0, batched_u_1)
+
+    v, g = value_and_grad(estimate_delta_f)(ref_params)
+
+    assert np.isfinite(v)
+    # assert v == ref_delta_f  # not expected in this case, due to non-physical fake_log_weights
+    assert np.isfinite(g).all()
+    assert (g != 0).any()
+    assert g.shape == ref_params.shape
+    # assert anything_about_direction_of_g  # not expected because the inputs are non-physical
+
+    # expect different estimate when evaluated on slightly different parameters
+    params_prime = ref_params + 0.01 * onp.random.randn(*ref_params.shape)
+    v_prime, g_prime = value_and_grad(estimate_delta_f)(params_prime)
+    assert v_prime != v
+    assert (g_prime != g).any()
+    assert np.isfinite(v_prime)
+    assert np.isfinite(g_prime).all()
+
+
+def test_one_sided_exp():
+    """assert consistency with pymbar.EXP on random instances"""
```

Comment:
Might be useful to add a test with `inf` values as well, I don't recall if pymbar does something clever to mask them out.

Reply:
will add a test of this

(logsumexp handles -inf gracefully, so this should be fine without additional trickery)

Reply:
ah good to know!

Reply:
Added in https://github.com/proteneer/timemachine/pull/654/commits/722a6d8eac7dfbe126b1f6be05e0015a13200cb7